<!DOCTYPE HTML PUBLIC "-//W3C//DTD HTML 4.01 Transitional//EN">
<html lang="en">
<head>
<link rel="stylesheet" href="stylesheets/styles.css">
<link rel="stylesheet" href="stylesheets/pygment_trac.css">
<meta name="viewport" content="width=device-width, initial-scale=1, user-scalable=no">

<meta http-equiv="Content-type" content="text/html; charset=UTF-8">
<meta http-equiv="Content-Style-Type" content="text/css">
<title>ASCI 994 / STAT 844</title>
</head>
<body>
<p class="menublock"><a href="http://morotalab.org/teaching/teaching.html">Back</a></p>

<h1>ASCI 944 / STAT 844 Quantitative Methods for Genomics of Complex Traits</h1>
<h2>Spring 2018</h2>


<br>
<br>
<h3>Instructor</h3>
<ul>
  <li>Name: Gota Morota</li>
  <li>Office: A218f Animal Science Building</li>
  <li>Email: morota@unl.edu</li>
  <li>Web: <a href="http://morotalab.org/">http://morotalab.org/</a></li>
  <li>Office Hours: By appointment</li>
</ul>

<br>
<br>
<h3>Time and Location</h3>
<ul>
  <li>Tues./Thurs. 9:30-10:45am</li>
  <li>Animal Science Building, Room A228</li>
</ul>

<br>
<br>
<h3>Prerequisites</h3>
<ul>
  <li>ASCI 861U, 931, or equivalent</li>
  <li>STAT 802, 821, or equivalent</li>
  <li>Knowledge of statistical programming language R</li>
  <li>Searle, S.R. (1982) <i>Matrix Algebra Useful for Statistics</i>. Wiley, New York. [<a href="http://www.amazon.com/Matrix-Algebra-Useful-Statistics-Shayle/dp/0470009616">Amazon</a>]</li>
</ul>





<br>
<br>
<h3>Course Description</h3>
This course will cover quantitative genetic analysis of complex trait genetics with emphasis on the use of molecular markers spanning the entire genome. We will discuss statistical methodologies for connecting phenotypes with high-dimensional genomic information to better understand polygenic traits from both prediction and inference perspectives. Topics will include genomic relatedness, linkage disequilibrium, population stratification, genomic heritability, missing heritability, genome-wide association study, genomic prediction, causal inference, and statistical learning. We will use examples from the animal, plant, and human genetics literature. Additional topics will be briefly touched upon, including sequence data, gene expression, epigenetics, and bioinformatics. Homework assignments involve hands-on analysis of simulated and real genomic data available at public repositories. The course will use R/Bioconductor software for statistical computing tools.
<br>
<br>
<h3>Learning Objectives</h3>
After taking this course, the student will be able to:
<ul>
  <li>understand the statistical theory behind commonly used quantitative methods in genomics</li>
  <li>apply statistical methods to high-dimensional genomic data and analyze them using statistical computing tools</li>
  <li>critically review current literature in statistical and quantitative genetics</li>
</ul>

<br>
<br>
<h3>Texts and Reading Materials</h3>
Lecture slides will be provided on the class website. There will be no required textbook.

<br>
<br>
<h3>Syllabus</h3>
<ul>
  <li>
    [<a href="asci944spring2018.pdf">PDF</a>]
  </li>
  </ul>
<br>
<br>

<h3>Schedule</h3>
Lectures will be delivered using a whiteboard and presentation slides. 
<ul>
  <li>1/9 (T): Course overview [<a href="http://morotalab.org/asci944-2018/day01/day01.html">HTML</a>]</li>
  <li>1/11 (R): Ordinary least-squares and the curse of dimensionality </li> 
  <li>1/16 (T): Variance, covariance, and linkage disequilibrium </li> 
  <li>1/18 (R): Covariance between allelic counts [<a href="http://morotalab.org/asci944-2018/hw1/hw1.html">HW1</a>]</li> 
  <li>1/23 (T): Review of average effect of allele substitution </li>
  <li>1/25 (R): Multi-locus additive genetic variance in the presence of linkage disequilibrium </li>
  <li>1/30 (T): Whole-genome regression - ridge regression  [<a href="http://morotalab.org/asci944-2018/day07/day07.html">R</a>] [<a href="http://morotalab.org/asci944-2018/hw2/hw2.html">HW2</a>]</li>
  <li>2/1 (R): Whole-genome regression - ridge regression  [<a href="http://morotalab.org/asci944-2018/day08/day08.html">R</a>] [<a href="http://dx.doi.org/10.1017/S0016672399004462">P</a>]</li> 
  <li>2/6 (T): Whole-genome regression - ridge regression</li>
  <li>2/8 (R): Overview of likelihood-based inference; Prediction of random effects - best prediction (BP) [<a href="http://morotalab.org/asci944-2018/day10/day10.html">R</a>][<a href="http://www.nasonline.org/publications/biographical-memoirs/memoir-pdfs/henderson-charles.pdf">P</a>]</li>
  <li>2/13 (T): Prediction of random effects - best linear prediction (BLP); best linear unbiased prediction (BLUP); mixed model equations (MME) (Guest lecture - Mr. Haipeng Yu) [<a href="https://haipengu.github.io/teaching/Tslides/BLUP_MME.pdf">PDF</a>][<a href="https://haipengu.github.io/teaching/Rcode/GBLUP.html">R</a>][<a href="https://www.jstor.org/stable/2245695">P</a>]</li>
  <li>2/15 (R): HW1 review & HW2 review (Guest lecture - Dr. Malachy Campbell)</li>
  <li>2/20 (T): Relatedness due to genetic markers - additive genomic relationship </li>
  <li>2/22 (R): Relatedness due to genetic markers - dominance genomic relationship </li>
  <li>2/27 (T): Relatedness due to genetic markers - dominance genomic relationship </li>
  <li>3/1 (R): Whole-genome regression - Genomic BLUP (GBLUP) [<a href="http://morotalab.org/asci944-2018/hw3/hw3.html">HW3</a>]</li>
  <li>3/6 (T):   </li>
  <li>3/8 (R): </li>
  <li>3/13 (T):  </li> 
  <li>3/15 (R):  </li> 
  <li>3/20 (T): </li> 
  <li>3/22 (R): </li> 
  <li>3/27 (T): </li>
  <li>3/29 (R): </li> 
  <li>4/3 (T):  </li> 
  <li>4/5 (R): </li> 
  <li>4/10 (T):  </li> 
  <li>4/12 (R):  </li>
  <li>4/17 (T):  </li>
  <li>4/19 (R):  </li>
  <li>4/24 (T):  </li>
  <li>4/26 (R):  </li>
  <li>5/1 (T):  </li>
</ul>






</body>
</html>
