<!DOCTYPE html>

<html xmlns="http://www.w3.org/1999/xhtml">

<head>

<meta charset="utf-8">
<meta http-equiv="Content-Type" content="text/html; charset=utf-8" />
<meta name="generator" content="pandoc" />

<meta name="author" content="Gota Morota" />


<title>ASCI 944 / STAT 844 Quantitative Methods for Genomics of Complex Traits</title>

<script src="hw3_files/jquery-1.11.3/jquery.min.js"></script>
<meta name="viewport" content="width=device-width, initial-scale=1" />
<link href="hw3_files/bootstrap-3.3.6/css/bootstrap.min.css" rel="stylesheet" />
<script src="hw3_files/bootstrap-3.3.6/js/bootstrap.min.js"></script>
<script src="hw3_files/jqueryui-1.11.4/jquery-ui.min.js"></script>
<script src="hw3_files/navigation-1.1/tabsets.js"></script>
<script src="hw3_files/navigation-1.1/codefolding.js"></script>
<link href="hw3_files/magnific-popup-1.1.0/magnific-popup.css" rel="stylesheet" />
<script src="hw3_files/magnific-popup-1.1.0/jquery.magnific-popup.min.js"></script>
<link href="hw3_files/readthedown-0.1/readthedown.css" rel="stylesheet" />
<script src="hw3_files/readthedown-0.1/readthedown.js"></script>


<style type="text/css">code{white-space: pre;}</style>
<style type="text/css">
div.sourceCode { overflow-x: auto; }
table.sourceCode, tr.sourceCode, td.lineNumbers, td.sourceCode {
  margin: 0; padding: 0; vertical-align: baseline; border: none; }
table.sourceCode { width: 100%; line-height: 100%; background-color: #ffffff; color: #1f1c1b; }
td.lineNumbers { text-align: right; padding-right: 4px; padding-left: 4px; background-color: #ffffff; color: #a0a0a0; border-right: 1px solid #a0a0a0; }
td.sourceCode { padding-left: 5px; }
pre, code { color: #1f1c1b; background-color: #ffffff; }
code > span.kw { color: #1f1c1b; font-weight: bold; } /* Keyword */
code > span.dt { color: #0057ae; } /* DataType */
code > span.dv { color: #b08000; } /* DecVal */
code > span.bn { color: #b08000; } /* BaseN */
code > span.fl { color: #b08000; } /* Float */
code > span.cn { color: #aa5500; } /* Constant */
code > span.ch { color: #924c9d; } /* Char */
code > span.sc { color: #3daee9; } /* SpecialChar */
code > span.st { color: #bf0303; } /* String */
code > span.vs { color: #bf0303; } /* VerbatimString */
code > span.ss { color: #ff5500; } /* SpecialString */
code > span.im { color: #ff5500; } /* Import */
code > span.co { color: #898887; } /* Comment */
code > span.do { color: #607880; } /* Documentation */
code > span.an { color: #ca60ca; } /* Annotation */
code > span.cv { color: #0095ff; } /* CommentVar */
code > span.ot { color: #006e28; } /* Other */
code > span.fu { color: #644a9b; } /* Function */
code > span.va { color: #0057ae; } /* Variable */
code > span.cf { color: #1f1c1b; font-weight: bold; } /* ControlFlow */
code > span.op { color: #1f1c1b; } /* Operator */
code > span.bu { color: #644a9b; font-weight: bold; } /* BuiltIn */
code > span.ex { color: #0095ff; font-weight: bold; } /* Extension */
code > span.pp { color: #006e28; } /* Preprocessor */
code > span.at { color: #0057ae; } /* Attribute */
code > span.re { color: #0057ae; } /* RegionMarker */
code > span.in { color: #b08000; } /* Information */
code > span.wa { color: #bf0303; } /* Warning */
code > span.al { color: #bf0303; font-weight: bold; } /* Alert */
code > span.er { color: #bf0303; text-decoration: underline; } /* Error */
code > span. { color: #1f1c1b; } /* Normal */
</style>


</head>

<body>


<div id="content" data-toggle="wy-nav-shift">

<!-- tabsets -->
<script>
$(document).ready(function () {
  window.buildTabsets("TOC");
});
</script>

<!-- code folding -->

<nav id="nav-top" role="navigation" aria-label="top navigation">
    <a role="button" href="#" data-toggle="wy-nav-top"><span class="glyphicon glyphicon-menu-hamburger"></span></a>
</nav>


<div id="header">
<h1 class="title">ASCI 944 / STAT 844 Quantitative Methods for Genomics of Complex Traits</h1>
<h1 class="subtitle">Homework assignment 3</h1>
</div>


<div id="table-of-contents">
    <h2><a href="#content">ASCI 944 / STAT 844 Quantitative Methods for Genomics of Complex Traits</a></h2>
    <div id="text-table-of-contents">
      <ul>
      <li><a href="#arabidopsis-data">Arabidopsis data</a><ul>
      <li><a href="#question-1">Question 1</a></li>
      <li><a href="#question-2">Question 2</a></li>
      <li><a href="#question-3">Question 3</a></li>
      <li><a href="#question-4">Question 4</a></li>
      <li><a href="#question-5">Question 5</a></li>
      <li><a href="#question-6">Question 6</a></li>
      <li><a href="#question-7">Question 7</a></li>
      <li><a href="#question-8">Question 8</a></li>
      </ul></li>
      <li><a href="#wheat-data">Wheat data</a><ul>
      <li><a href="#question-9">Question 9</a></li>
      </ul></li>
      <li><a href="#mouse-data">Mouse data</a><ul>
      <li><a href="#question-10">Question 10</a></li>
      </ul></li>
      <li><a href="#maize-data">Maize data</a><ul>
      <li><a href="#question-11">Question 11</a></li>
      <li><a href="#question-12">Question 12</a></li>
      </ul></li>
      </ul>
    </div>
</div>

<div id="main">
<div id="due-date" class="section level3">
<h3>Due date</h3>
<p>Tuesday, March 13, 5pm</p>
</div>
<div id="arabidopsis-data" class="section level1">
<h1>Arabidopsis data</h1>
<p>We will analyze Arabidopsis recombinant inbred lines available at the <a href="http://publiclines.versailles.inra.fr/page/33">Bay-0 x Shahdara</a> project. Download the <a href="http://publiclines.versailles.inra.fr/ijpbinfo/publiclines/33/BayxSha_PublishedPheno.xls">phenotypic</a> and <a href="http://publiclines.versailles.inra.fr/ijpbinfo/publiclines/33/BayxSha_2_Genotypes.xls">genotypic</a> data. The target trait in this homework assignment is <span style="color:red"> Flowering time in Long Days</span> measured as the number of days between germination and bolting (the beginning of the inflorescence growth). You can learn more about the Bay-0 x Shahdara population in <a href="http://dx.doi.org/10.1007/s00122-001-0825-9">Loudet et al., (2002)</a>.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(readxl)
<span class="co"># phenotypes</span>
bayxshaY &lt;-<span class="st"> </span>(<span class="kw">read_excel</span>(<span class="st">&quot;BayxSha_PublishedPheno.xls&quot;</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>, <span class="dt">sheet =</span> <span class="dv">1</span>, 
    <span class="dt">skip =</span> <span class="dv">4</span>))
y0 &lt;-<span class="st"> </span><span class="kw">as.numeric</span>(<span class="kw">unlist</span>(bayxshaY[, <span class="dv">6</span>]))
y1 &lt;-<span class="st"> </span>y0[<span class="op">-</span><span class="kw">which</span>(<span class="kw">is.na</span>(y0))]
y1 &lt;-<span class="st"> </span><span class="kw">scale</span>(y1)

<span class="co"># genotypes</span>
bayxshaG &lt;-<span class="st"> </span><span class="kw">read_excel</span>(<span class="st">&quot;BayxSha_2_Genotypes.xls&quot;</span>, <span class="dt">col_names =</span> <span class="ot">FALSE</span>, <span class="dt">skip =</span> <span class="dv">5</span>)
bayxshaG.id &lt;-<span class="st"> </span>bayxshaG[, <span class="dv">1</span>]
bayxshaG &lt;-<span class="st"> </span>bayxshaG[, <span class="op">-</span><span class="dv">1</span>]
bayxshaG[bayxshaG <span class="op">==</span><span class="st"> &quot;A&quot;</span>] &lt;-<span class="st"> </span><span class="dv">2</span>
bayxshaG[bayxshaG <span class="op">==</span><span class="st"> &quot;C&quot;</span>] &lt;-<span class="st"> </span><span class="dv">1</span>
bayxshaG[bayxshaG <span class="op">==</span><span class="st"> &quot;B&quot;</span>] &lt;-<span class="st"> </span><span class="dv">0</span>
bayxshaG[bayxshaG <span class="op">==</span><span class="st"> &quot;D&quot;</span>] &lt;-<span class="st"> </span><span class="ot">NA</span>
bayxshaG &lt;-<span class="st"> </span><span class="kw">as.matrix</span>(bayxshaG)
<span class="kw">mode</span>(bayxshaG) &lt;-<span class="st"> &quot;numeric&quot;</span>
bayxshaG &lt;-<span class="st"> </span>bayxshaG[<span class="op">-</span><span class="kw">which</span>(<span class="kw">is.na</span>(y0)), ]</code></pre></div>
<div id="question-1" class="section level2">
<h2>Question 1</h2>
<p>What are the dimensions of <code>y1</code> and <code>bayxshaG</code>?</p>
</div>
<div id="question-2" class="section level2">
<h2>Question 2</h2>
<p>How many markers contain missing genotypes (<code>NA</code>) ? Use the function <code>is.na()</code>.</p>
</div>
<div id="question-3" class="section level2">
<h2>Question 3</h2>
<p>Replace missing marker genotypes with mean values in a column-wise fashion (i.e., compute the mean of each column and replace NAs in that column). Then store marker genotypes in a matrix variable <code>W1</code>. Show that the variable <code>W1</code> does not include any <code>NA</code>.</p>
</div>
<div id="question-4" class="section level2">
<h2>Question 4</h2>
<p>Apply a singular value decomposition (SVD) to the genotype matrix <code>W1</code> (i.e., <span class="math inline">\(\mathbf{W}_1 = \mathbf{UDV&#39;}\)</span>). Show that 1) <span class="math inline">\(\mathbf{U&#39;U}= \mathbf{I}\)</span>, 2) <span class="math inline">\(\mathbf{V&#39;V} = \mathbf{I}\)</span>, 3) <span class="math inline">\(\mathbf{VV}&#39;= \mathbf{I}\)</span>, 4) <span class="math inline">\(\mathbf{V^{-1}} = \mathbf{V&#39;}\)</span>, and 5) <span class="math inline">\(\mathbf{V&#39;}^{-1} = \mathbf{V}\)</span>. Use the function <code>svd()</code>.</p>
</div>
<div id="question-5" class="section level2">
<h2>Question 5</h2>
<p>Fit a standard OLS and a principal component-based OLS. Show that the regression of phenotypes on marker covariates is equivalent to the regression of phenotypes on principal components (eigenvectors) computed from marker genotypes. Hint: One way to answer this question is to show that their predicted genetic values are equal.</p>
</div>
<div id="question-6" class="section level2">
<h2>Question 6</h2>
<p>In general, the ridge regression has a higher predictive ability than OLS because the vector of ridge estimators yields a smaller variance than that of the vector of OLS estimators. Fit a standard ridge regression. Use <span class="math inline">\(\lambda = 5\)</span>. Show that the variance of marker effect estimates is smaller in the ridge regression than that of OLS.</p>
</div>
<div id="question-7" class="section level2">
<h2>Question 7</h2>
<p>Fit a standard ridge regression and a SVD-based ridge regression. Use <span class="math inline">\(\lambda = 5\)</span>. Show that their predicted genetic values are equal.</p>
</div>
<div id="question-8" class="section level2">
<h2>Question 8</h2>
<p>Recall that the effect of ridge is to replace <span class="math inline">\(D^{-1}\)</span> in the OLS expression with <span class="math inline">\((D^2 + \lambda I)^{-1}D\)</span>. Show that ridge estimates are shrunk toward zero because these amounts are indeed smaller in the ridge regression than those in OLS. Use <span class="math inline">\(\lambda = 5\)</span>.</p>
</div>
</div>
<div id="wheat-data" class="section level1">
<h1>Wheat data</h1>
<p>We will analyze the wheat data from Pérez-Rodríguez et al., (2012). (<a href="http://www.g3journal.org/content/2/12/1595">doi:10.1534/g3.112.003665</a>). Download the supporting data <a href="http://www.g3journal.org/content/2/12/1595.supplemental">File S1</a>. The R object <code>YLD_1.RData</code> includes 306 elite wheat lines genotyped with 1,717 diversity array technology (DArT) markers. Note that the <a href="https://en.wikipedia.org/wiki/Diversity_Arrays_Technology">DArT marker</a> is coded either 0 or 1 indicating the presence or absence of allele at a given locus. The marker information is in the <code>Marker</code> variable.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())
<span class="kw">load</span>(<span class="st">&quot;data_G3/YLD/YLD_1.RData&quot;</span>)
W2 &lt;-<span class="st"> </span>Markers
<span class="kw">dim</span>(W2)</code></pre></div>
<div id="question-9" class="section level2">
<h2>Question 9</h2>
<p>Recall that ridge estimates can be expressed as <span class="math inline">\(\hat{\beta}^{ridge}_j = \frac{\sum^n_{i=1} x^2_{ij} }{\sum^n_{i=1} x^2_{ij} + \lambda} \times \hat{\beta}^{ols}_{j}\)</span>. Create a centered genotype matrix <code>W2c</code>. Calculate the minor allele frequencies (MAF) and the amounts of shrinkage for all markers. Plot MAF on the X axis and the amounts of shrinkage on the Y axis. Interpret the figure you obtained. Use <span class="math inline">\(\lambda = 100\)</span>.</p>
</div>
</div>
<div id="mouse-data" class="section level1">
<h1>Mouse data</h1>
<p>We will analyze the mouse data from Valdar et al., (2006) (<a href="http://www.nature.com/ng/journal/v38/n8/abs/ng1840.html">doi:10.1038/ng1840</a>). This dataset is available in the <a href="https://cran.r-project.org/web/packages/BGLR/index.html">BGLR</a> R package. The <code>mice.X</code> variable contains 1,814 individuals and 10,346 SNPs.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">library</span>(BGLR)
<span class="kw">data</span>(mice)
<span class="st">`</span><span class="dt">?</span><span class="st">`</span>(mice)
W3 &lt;-<span class="st"> </span>mice.X
<span class="kw">dim</span>(W3)</code></pre></div>
<div id="question-10" class="section level2">
<h2>Question 10</h2>
<p>Calculate MAF and the amounts of shrinkage for all markers. Create a centered genotype matrix <code>W3c</code>. Plot MAF on the X axis and the amounts of shrinkage on the Y axis. Interpret the figure you obtained. Does the pattern you observe here agree with the wheat dataset from Question 9? If not, explain why. Use <span class="math inline">\(\lambda = 100\)</span>.</p>
</div>
</div>
<div id="maize-data" class="section level1">
<h1>Maize data</h1>
<p>Download the maize data <em>File S1</em> from Crossa et al., (2010) available at <a href="http://dx.doi.org/10.1534/genetics.110.118521">10.1534/genetics.110.118521</a>. The <code>dataCorn_SS.RData</code> contains 264 corn yield phenotypes evaluated under severe drought stress conditions. The variable <code>X</code> includes 1,135 SNP markers.</p>
<div class="sourceCode"><pre class="sourceCode r"><code class="sourceCode r"><span class="kw">rm</span>(<span class="dt">list =</span> <span class="kw">ls</span>())
<span class="kw">load</span>(<span class="st">&quot;analysis_wheat_corn/corn_yield/data/dataCorn_SS.RData&quot;</span>)
W4 &lt;-<span class="st"> </span>X
<span class="kw">dim</span>(W4)</code></pre></div>
<div id="question-11" class="section level2">
<h2>Question 11</h2>
<p>Calculate MAF and the amounts of shrinkage for all markers. Create a centered genotype matrix <code>W4c</code>. Plot MAF on the X axis and the amounts of shrinkage on the Y axis. Interpret the figure you obtained. Does the pattern you observe agree with the wheat dataset from Question 9? If not, explain why. Use <span class="math inline">\(\lambda = 100\)</span>.</p>
</div>
<div id="question-12" class="section level2">
<h2>Question 12</h2>
<p>Continue analyzing the maize dataset. The vector of phenotypes is included in the <code>y</code> variable. Recall that the dimension of predictors is 264 by 1,135. Show that we cannot fit OLS using the <code>solve()</code> function when <span class="math inline">\(n &lt; m\)</span> but the ridge regression can circumvent this problem. Create a scatter plot of observed <code>y</code> vs. predicted <code>y</code>. Use the variables <code>y</code> and <code>W4</code>. Set <span class="math inline">\(\lambda = 100\)</span>.</p>
</div>
</div>
</div>


</div>

<div id="postamble" data-toggle="wy-nav-shift" class="status">
<p class="author"><span class="glyphicon glyphicon-user"></span> Gota Morota</p>
<p class="date"><span class="glyphicon glyphicon-calendar"></span> March 1, 2018</p>
</div>


<script>
$(document).ready(function () {
 	  });
</script>


<!-- dynamically load mathjax for compatibility with self-contained -->
<script>
  (function () {
    var script = document.createElement("script");
    script.type = "text/javascript";
    script.src  = "https://cdnjs.cloudflare.com/ajax/libs/mathjax/2.7.1/MathJax.js?config=TeX-AMS-MML_HTMLorMML";
    document.getElementsByTagName("head")[0].appendChild(script);
  })();
</script>

</body>
</html>
